lib/tracing-limit/src/lib.rs

```diff
@@ -10,199 +9,204 @@ use std::{
-                                &message
-                            } else {
-                                "unknown event"
-                            };
-
-                            let meta = event.metadata();
-
-                            println!(
-                                "{} {} Limiting the output of \"{}\" for the next {} seconds",
-                                FmtLevel(&tracing_core::Level::INFO),
-                                meta.target(),
-                                message,
-                                limit
-                            );
-                            return;
-                        } else if local_count > 1 {
-                            return;
-                        }
-                    } else {
-                        drop(events);
-
-                        let mut events = self.events.write().unwrap();
-                        events.remove(&id);
-                        drop(events);
-
-                        let meta = event.metadata();
-
-                        let mut visitor = FmtVisitor::default();
-                        event.record(&mut visitor);
-
-                        let message = if let Some(message) = &visitor.message {
-                            &message
+    fn enabled(&self, metadata: &Metadata, ctx: Context<S>) -> bool {
+        if is_limited(metadata) {
+            let id = metadata.callsite();
+            let events = self.events.read().expect("lock poisoned!");
+
+            // check if the event exists within the map, if it does
+            // that means we are currently rate limiting it.
+            if let Some(state) = events.get(&id) {
+                let ts = SystemTime::now()
+                    .duration_since(UNIX_EPOCH)
+                    .unwrap()
+                    .as_secs() as usize;
+
+                if ts - state.start < state.limit {
+                    if state.count.load(Ordering::SeqCst) == 1 {
+                        let message = if let Some(event_message) = &state.message {
+                            format!("{:?} is being rate limited.", event_message)
                         } else {
-                            "unknown event"
+                            format!("unknown message is being rate limited.")
                         };
 
-                        // We need to replicate the way that fmt logs events
-                        // because we currently can not create new fresh events.
-                        println!(
-                            "{} {} {:?} {:?} logs were rate limited.",
-                            FmtLevel(meta.level()),
-                            meta.target(),
-                            local_count,
-                            message
-                        );
-
-                        return;
+                        self.create_event(&id, &ctx, message);
+                    }
+
+                    let prev = state.count.fetch_add(1, Ordering::SeqCst);
```

Comment:
As I mentioned above, this doesn't need to be sequentially consistent:
```suggestion
                    let prev = state.count.fetch_add(1, Ordering::Release);
```